### PR TITLE
Add depends-ciphers.pl

### PR DESCRIPTION
**Pull request by mpg**
## Description

Add a new test script `depends-ciphers.pl` - this is a step towards https://github.com/ARMmbed/mbedtls/issues/1892

## Status
**READY**

## Requires Backporting
YES

- This PR can be merged both in development and 2.16
- Backport to 2.7: #2316 

## Migrations
NO

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated
- [x] Backported

### CODE REVIEW

File: tests/scripts/depends-ciphers.pl [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# Some algorithms can't be disabled on their own as others depend on them, so
+# we list those reverse-dependencies here to keep check_config.h happy.
+my %revdeps = (
+    'MBEDTLS_AES_C'         => ['MBEDTLS_CTR_DRBG_C'],
+    'MBEDTLS_CHACHA20_C'    => ['MBEDTLS_CHACHAPOLY_C'],
+);
+
+system( "cp $config_h $config_h.bak" ) and die;
+sub abort {
+    #system( "mv $config_h.bak $config_h" ) and warn "$config_h not restored\n";
```
[the remark this answers is not in the document]
Comment (mpg) [reply tests/scripts/depends-ciphers.pl]: Good catch, it's a leftover from debugging test failures.

**Comment by mpg** [comment]
@sbutcher-arm @gilles-peskine-arm Thanks for your review! I believe I've addressed your comments, please review again.

File: tests/scripts/depends-ciphers.pl [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+# get a list of ciphers from the cipher module
+my $cipher_h = 'include/mbedtls/cipher.h';
+my $sed_cmd = 's/^    MBEDTLS_CIPHER_ID_\([A-Z0-9_]*\),.*/\1/p';
+my @ciphers = split( /\s+/, `sed -n -e '$sed_cmd' $cipher_h` );
+@ciphers = map { "MBEDTLS_${_}_C" } grep { ! m/NULL|3DES/ } @ciphers;
+
+# Some algorithms can't be disabled on their own as others depend on them, so
+# we list those reverse-dependencies here to keep check_config.h happy.
+my %revdeps = (
+    'MBEDTLS_AES_C'         => ['MBEDTLS_CTR_DRBG_C'],
```
Comment (RonEld) [inline tests/scripts/depends-ciphers.pl]: There are other dependencies on `MBEDTLS_AES_C`:
- `MBEDTLS_CMAC_C` ( also dependent on `MBEDTLS_DES_C`)
- `MBEDTLS_NIST_KW_C` 
- `MBEDTLS_GCM_C` ( also dependent on `MBEDTLS_CAMELLIA_C`)

We shouldn't rely on the default configuration.

Comment (RonEld) [reply tests/scripts/depends-ciphers.pl]: That's what I thought, and this is why I said that we shouldn't rely on the default configuration.

What if B is already undefined in the original config?

how about we start with a full configuration, and disable one by one?

Comment (mpg) [reply tests/scripts/depends-ciphers.pl]: I agree that we should start from a full config.

Comment (sbutcher-arm) [reply tests/scripts/depends-ciphers.pl]: Am I missing something, or is it still starting from the default config?

I don't mind the script dealing with whatever `config.h` it's given, but the `all.sh` run seems to be the default.

Comment (mpg) [reply tests/scripts/depends-ciphers.pl]: I don't think you're missing anything. This part is pending rework.

File: tests/suites/test_suite_cipher.function [hunk 3]
```diff
@@ -306,7 +306,7 @@ exit:
 }
 /* END_CASE */
 
-/* BEGIN_CASE */
+/* BEGIN_CASE depends:MBEDTLS_AES_C:MBEDTLS_CIPHER_MODE_CBC */
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_cipher.function]: Typo: `depends:` → `depends_on:`

**Comment by mpg** [comment]
retest

**Comment by mpg** [comment]
@sbutcher-arm @RonEld @gilles-peskine-arm I believed I addressed your feedback. As the review history wasn't too significant, I've taken the liberty of rewriting the history in order to resolve conflicts (in particular with #930) by rebasing on current development and more cleanly separate fixes to test cases from changes in the script (in the direction suggest by Simon).

The main change is the script now starts from a full configuration as suggested by all of you. As a result, I've found and fixed errors in the CMAC test suite. I've been tempted to do the same in other `depends-xxx.pl` scripts but the first attempt found non-trivial issues, so I think it would better be handled in a separate PR in order to keep PRs nice and small.

Please (at least two of you) review again.

**Comment by Patater** [comment]
This PR cannot run CI due to it being based on a branch that doesn't contain a Jenkinsfile. Will run CI manually as part of gatekeeping.

**Comment by mpg** [comment]
I've just rebased on current development in order to resolve conflicts (and as a welcome side-effect, enable Jenkins testing). @Patater @gilles-peskine-arm please review again. @RonEld @sbutcher-arm you previously requested changes and I think I did all of them, please check and update your review status.

**Comment by gilles-peskine-arm** [comment]
Please make a PR for mbed-crypto once this one is ready.

**Comment by mpg** [comment]
This PR would require significant rework, and a consolidated dependency checking script is in the pipeline, so it's probably best to focus our efforts on the more general solution rather than on this PR.

**Pull request closed without merging**